engine/src/main/java/org/entando/entando/aps/system/services/digitalexchange/client/DigitalExchangesClientImpl.java

```diff
@@ -0,0 +1,181 @@
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Component;
+import org.springframework.web.client.RestClientException;
+import org.springframework.web.client.RestClientResponseException;
+import org.springframework.web.client.RestTemplate;
+
+/**
+ * Implementation notes. We want to be able to configure a possibly different
+ * timeout for each DE instance. Currently this implementation uses the Spring
+ * RestTemplate class for performing HTTP calls. RestTemplate doesn't provide a
+ * way for configuring a different timeout for each call (it can only be
+ * configured globally), however it is possible to define a timeout using the
+ * CompletableFuture class. Currently it is necessary to use CompletableFuture
+ * combined with a ScheduledExecutorService to achieve this. Starting from Java
+ * 9 it will be possible to simplify the code using the new orTimeout() method
+ * of CompletableFuture. Moreover the RestTemplate class could be deprecated in
+ * the future in favor of the new WebClient (this will imply adding Spring
+ * WebFlux dependency to the project).
+ */
+@Component
+public class DigitalExchangesClientImpl implements DigitalExchangesClient {
+
+    private static final Logger logger = LoggerFactory.getLogger(DigitalExchangesClientImpl.class);
+
+    public static final String ERRCODE_DE_HTTP_ERROR = "1";
+    public static final String ERRCODE_DE_UNREACHABLE = "2";
+    public static final String ERRCODE_DE_TIMEOUT = "3";
+    public static final String ERRCODE_DE_INVALID_URL = "4";
+    public static final String ERRCODE_DE_WRONG_PAYLOAD = "5";
+
+    private static final int DEFAULT_TIMEOUT = 10000;
+
+    private final DigitalExchangesService digitalExchangesService;
+    private final RestTemplate restTemplate;
+    private final MessageSource messageSource;
+    private final ScheduledExecutorService timeoutService;
+
+    @Autowired
+    public DigitalExchangesClientImpl(DigitalExchangesService digitalExchangesService,
+            RestTemplate restTemplate, MessageSource messageSource) {
+        this.digitalExchangesService = digitalExchangesService;
+        this.restTemplate = restTemplate;
+        this.messageSource = messageSource;
+        timeoutService = Executors.newSingleThreadScheduledExecutor();
+    }
+
+    @Override
+    public <R extends RestResponse, C> C getCombinedResult(DigitalExchangeCall<R, C> call) {
+        List<R> allResults = queryAllDigitalExchanges(call);
+        return call.combineResults(allResults);
+    }
+
+    private <R extends RestResponse, C> List<R> queryAllDigitalExchanges(DigitalExchangeCall<R, C> call) {
+
+        CompletableFuture<R>[] futureResults
+                = digitalExchangesService.getDigitalExchanges()
+                        .stream()
+                        .filter(de -> de.isActive())
+                        .map(de -> getResponseOrTimeout(de, call))
+                        .toArray(CompletableFuture[]::new);
```

Comment:
In this case that we can't get rid of the unchecked warning I would add a  @SuppressWarnings("unchecked") and a comment saying that it's OK because java can't resolve generics here